PR: https://github.com/odoo/odoo/pull/215517

From: e122564739ce061b935d3289d72444fe0595bb16
From: chha-odoo
Date: 2025-09-07 06:44:57

Breaking data model changes scores: del:2 + add:2, change matches:
-    _inherit = 'stock.location'
-    is_a_dock = fields.Boolean("Is a Dock Location")
+    dock_ids = fields.Many2many(
+    allowed_dock_ids = fields.Many2many(related='picking_type_id.dock_ids', string="Allowed Docks")

Total Changes: 40

[IMP] stock_fleet, stock: UI Improvements for dock location

Some dock locations are used for only delivery purpose, some used for
only reciept purpose. so we are making it operation type specific.
so user can set the dock locations from the operation type form view
and use this locations while creating batch.

With this commit:
*----------------------------------------------
* When `Dispatch Management` and `Storage Locations` are enabled, users can
  select dock locations by `To field` from the `Operation Type form view`.
* In the `picking batch form view`, only the selected dock locations and their
  child locations (defined in the operation type) are shown in the dropdown.
* `WH/Output` is set as the default dock location for `Delivery operation type`.
  as we commonly used this location for delivery.
* Formatted display name for locations in `M2M` and `M2O` fields as
  `parent location (muted) / current location (bold)` which helps users
  to easily identify the location within its hierarchy.

as there are no fixed dock locations now.
* Removed the `Is a Dock Location` field from the location form/tree view.
* Removed the demo data for the dock locations (Dock A and Dock B)

Task ID: [4728397](https://www.odoo.com/odoo/project/966/tasks/4728397)

Part-of: odoo/odoo#215517
Related: odoo/enterprise#88319
Related: odoo/upgrade#7991
Signed-off-by: Arnold Moyaux (arm) <arm@odoo.com>

================================= pseudo patch: =================================

--- a/addons/stock_fleet/models/__init__.py
+++ b/addons/stock_fleet/models/__init__.py
@@ -1,5 +1,4 @@
 from . import fleet_vehicle_model
 from . import stock_picking_batch
 from . import stock_picking
-from . import stock_location
 from . import stock_warehouse

--- a/addons/stock_fleet/models/stock_location.py
+++ b/None
@@ -1,7 +0,0 @@
-from odoo import fields, models
-
-
-class StockLocation(models.Model):
-    _inherit = 'stock.location'
-
-    is_a_dock = fields.Boolean("Is a Dock Location")

--- a/addons/stock_fleet/models/stock_picking.py
+++ b/addons/stock_fleet/models/stock_picking.py
@@ -1,6 +1,6 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import fields, models
+from odoo import Command, api, fields, models
 
 
 class StockPickingType(models.Model):
@@ -10,6 +10,18 @@ class StockPickingType(models.Model):
         'Dispatch Management',
         help="Enable this option to display dispatch management related details in the batch/wave form view and operations kanban overview."
     )
+    dock_ids = fields.Many2many(
+        'stock.location',
+        'dock_location_stock_picking_type_rel',
+        domain="[('warehouse_id', '=', warehouse_id), ('usage', '=', 'internal')]",
+        compute='_compute_dock_ids', store=True, readonly=False
+    )
+
+    @api.depends('warehouse_id')
+    def _compute_dock_ids(self):
+        for picking_type in self:
+            if picking_type.warehouse_id != picking_type._origin.warehouse_id and picking_type.dock_ids:
+                picking_type.dock_ids = [Command.clear()]
 
 
 class StockPicking(models.Model):

--- a/addons/stock_fleet/models/stock_picking_batch.py
+++ b/addons/stock_fleet/models/stock_picking_batch.py
@@ -12,8 +12,9 @@ class StockPickingBatch(models.Model):
     vehicle_category_id = fields.Many2one(
         'fleet.vehicle.model.category', string="Vehicle Category",
         compute='_compute_vehicle_category_id', store=True, readonly=False)
-    dock_id = fields.Many2one('stock.location', string="Dock Location", domain="[('warehouse_id', '=', warehouse_id), ('is_a_dock', '=', True)]",
-                              compute='_compute_dock_id', store=True, readonly=False)
+    allowed_dock_ids = fields.Many2many(related='picking_type_id.dock_ids', string="Allowed Docks")
+    dock_id = fields.Many2one('stock.location', string="Dock", domain="[('id', 'child_of', allowed_dock_ids)]",
+        compute='_compute_dock_id', store=True, readonly=False)
     vehicle_weight_capacity = fields.Float(string="Vehcilce Payload Capacity",
                               related='vehicle_category_id.weight_capacity')
     weight_uom_name = fields.Char(string='Weight unit of measure label', compute='_compute_weight_uom_name')
@@ -41,12 +42,13 @@ class StockPickingBatch(models.Model):
         for rec in self:
             rec.vehicle_category_id = rec.vehicle_id.category_id
 
-    @api.depends('picking_ids', 'picking_ids.location_id', 'picking_ids.location_dest_id')
+    @api.depends('picking_ids', 'picking_ids.location_id', 'picking_ids.location_dest_id', 'picking_type_id')
     def _compute_dock_id(self):
         for batch in self:
-            if batch.picking_ids:
-                if len(batch.picking_ids.location_id) == 1 and batch.picking_ids.location_id.is_a_dock:
-                    batch.dock_id = batch.picking_ids.location_id
+            if batch.picking_type_id != batch._origin.picking_type_id and batch.dock_id:
+                batch.dock_id = False
+            if batch.picking_ids and len(batch.picking_ids.location_id) == 1 and batch.picking_ids.location_id in batch.allowed_dock_ids:
+                batch.dock_id = batch.picking_ids.location_id
 
     def _compute_weight_uom_name(self):
         self.weight_uom_name = self.env['product.template']._get_weight_uom_name_from_ir_config_parameter()

--- a/addons/stock_fleet/models/stock_warehouse.py
+++ b/addons/stock_fleet/models/stock_warehouse.py
@@ -1,4 +1,4 @@
-from odoo import models
+from odoo import Command, models
 
 
 class StockWarehouse(models.Model):
@@ -15,6 +15,8 @@ class StockWarehouse(models.Model):
 
         if values.get('out_type_id'):
             values['out_type_id']['dispatch_management'] = True
+            if self.delivery_steps in ('pick_ship', 'pick_pack_ship'):
+                values['out_type_id']['dock_ids'] = [Command.link(self.wh_output_stock_loc_id.id)]
         if values.get('in_type_id'):
             values['in_type_id']['dispatch_management'] = True
 
